### ISSUE TITLE
can't use MapBase in SDK

### ISSUE THREAD

**Issue by jmesserly** [dart-lang/dev_compiler#344]
This is happening in the YAML package.

containsValue in MapMixin appears to be incompatible with the one in Map (see https://api.dartlang.org/1.12.1/dart-collection/MapMixin/containsValue.html vs https://api.dartlang.org/1.12.1/dart-core/Map/containsValue.html).

MapMixin has `bool containsValue(V value)` and Map has `bool containsValue(Object value)`

it produces:

```
[error] Base class introduces an invalid override. The type of MapMixin.containsValue ((dynamic) → bool) is not a subtype of Map<dynamic, dynamic>.containsValue ((Object) → bool). (yaml/lib/src/yaml_node_wrapper.dart, line 17, col 22)
```

I'm confused on a few levels...

Is this a type error with the SDK, that just happens to show up in package:yaml because it's using a mixin? Or maybe because it's substituting dynamic as V?

I'm also wondering if this needs to be an error when the type of the parameter depends on a generic type. Because of covariance, we'd have to check the parameter's type at runtime anyway. In other words, we already allow converting `MapMixin<V>` to `MapMixin<Object>` for any V. So the override is sound at runtime already. That said, it may still be a helpful error: if folks expect to be able to pass any Object to Map.containsValue and it throws in some common cases like MapMixin, that's pretty inconsistent.

Anyway, maybe the bug is just in the SDK, but something is wrong here, because you can't use dart:collection MapBase without an error from DDC.

**Comment by a14n** [comment]
It looks like an issue with the SDK (and the analyzer - not in strong mode). `MapMixin<K, V>` does not implements the method `containsValue(Object o)` because `MapMixin<K,int>` will not accept `containsValue("")` that is allowed by `Map<K, int>`.

Playing a little more with generics and `implements`/`extends` it's strange how generics are currently handled. I just filed dart-lang/sdk#24413

**Comment by lrhn** [comment]
This is a bug in the `MapMixin` class. Its `containsValue` should take `Object` as argument like the `Map` interface says. I'll fix it.

**Comment by a14n** [comment]
@lrhn see https://github.com/dart-lang/sdk/pull/24414 :)

**Comment by lrhn** [comment]
I've landed a fix including the on in dart-lang/sdk#24414 and a few other cases of the same problem (and some more clean-up while I was there).

**Comment by jmesserly** [comment]
Wow, awesome! Thank you!
